File: frontend-api/src/main/java/no/nav/melosys/tjenester/gui/dto/utpeking/UtpekingsperiodeDto.java
package no.nav.melosys.tjenester.gui.dto.utpeking;

import java.time.LocalDate;

import no.nav.melosys.domain.Utpekingsperiode;
import no.nav.melosys.domain.jpa.LovvalgBestemmelsekonverterer;
import no.nav.melosys.domain.kodeverk.Land_iso2;

public record UtpekingsperiodeDto(LocalDate fomDato,
                                  LocalDate tomDato,
                                  String lovvalgsbestemmelse,
                                  String tilleggsbestemmelse,
                                  String lovvalgsland) {
    private static final LovvalgBestemmelsekonverterer konverterer = new LovvalgBestemmelsekonverterer();

    public static UtpekingsperiodeDto av(Utpekingsperiode utpekingsperiode) {
        return new UtpekingsperiodeDto(
            utpekingsperiode.getFom(),
            utpekingsperiode.getTom(),
            utpekingsperiode.getBestemmelse() != null ? utpekingsperiode.getBestemmelse().name() : null,
            utpekingsperiode.getTilleggsbestemmelse() != null ? utpekingsperiode.getTilleggsbestemmelse().name() : null,
            utpekingsperiode.getLovvalgsland() != null ? utpekingsperiode.getLovvalgsland().name() : null
        );
    }

    public final Utpekingsperiode tilDomene() {
        return new Utpekingsperiode(
            fomDato,
            tomDato,
            enumVerdiEllerNull(Land_iso2.class, lovvalgsland),
            konverterer.convertToEntityAttribute(lovvalgsbestemmelse),
            konverterer.convertToEntityAttribute(tilleggsbestemmelse));
    }

    private static <E extends Enum<E>> E enumVerdiEllerNull(Class<E> enumKlasse, String nøkkel) {
        return nøkkel == null ? null : Enum.valueOf(enumKlasse, nøkkel);
    }
}

File: frontend-api/src/main/kotlin/no/nav/melosys/tjenester/gui/dto/utpeking/UtpekingsperiodeDto.kt
package no.nav.melosys.tjenester.gui.dto.utpeking;
import no.nav.melosys.domain.Utpekingsperiode
import no.nav.melosys.domain.jpa.LovvalgBestemmelsekonverterer
import no.nav.melosys.domain.kodeverk.Land_iso2
import java.time.LocalDate

data class UtpekingsperiodeDto(
    val fomDato: LocalDate,
    val tomDato: LocalDate,
    val lovvalgsbestemmelse: String,
    val tilleggsbestemmelse: String?,
    val lovvalgsland: String
) {
    companion object {
        private val konverterer = LovvalgBestemmelsekonverterer()

        fun av(utpekingsperiode: Utpekingsperiode): UtpekingsperiodeDto {
            return UtpekingsperiodeDto(
                utpekingsperiode.fom,
                utpekingsperiode.tom,
                utpekingsperiode.bestemmelse.name(),
                utpekingsperiode.tilleggsbestemmelse?.name(),
            utpekingsperiode.lovvalgsland.name
            )
        }
    }

    fun tilDomene(): Utpekingsperiode {
        return Utpekingsperiode(
            fomDato,
            tomDato,
            enumVerdiEllerNull(Land_iso2::class.java, lovvalgsland),
        konverterer.convertToEntityAttribute(lovvalgsbestemmelse),
            konverterer.convertToEntityAttribute(tilleggsbestemmelse)
        )
    }

    private fun <E : Enum<E>> enumVerdiEllerNull(enumKlasse: Class<E>, nøkkel: String?): E? {
        return nøkkel?.let {
            java.lang.Enum.valueOf(enumKlasse, it)
        }
    }
}
